src/client/workflow/workflow.ts

```diff
@@ -0,0 +1,262 @@
+          (step, stepOther) => step.stepId - stepOther.stepId
+        );
+        const concurrentResults = sortedSteps
+          .filter((step) => step.stepId >= this.stepCount)
+          .map((step) => step.out)
+          .slice(0, stepFunctions.length) as TResults;
+        this.stepCount += stepFunctions.length - 1;
+        this.planStepCount += stepFunctions.length;
+        return concurrentResults;
+      }
+    }
+    await this.sendPendingToQstash();
+    this.skip = true;
+    return [] as unknown as TResults;
+  }
+
+  /**
+   * Determines the parallel call state
+   *
+   * Parallel can be called in three states:
+   * 1. Called for the first time: will send the three plans to qstash one by one
+   * 2. Called with partial results: after the initial call, will be called multiple
+   *    times with each call having a unique targetStep. Corresponding target step
+   *    will be executed
+   * 3. Called with full results: After the final partial result call returns to Qstash,
+   *    qstash will call again with the full result. In this case, parallel step will
+   *    return the result and
+   */
+  protected getParallelCallState(parallelStepCount: number): PARALLEL_CALL_STATE {
+    const steps = this.steps.filter(
+      (step) => (step.stepId === 0 ? step.targetStep : step.stepId) >= this.stepCount
+    );
+
+    if (steps.length === 0) {
+      return "first";
+      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
+    } else if (steps.length >= 2 * parallelStepCount) {
+      return "last";
+    } else if (steps.at(-1)?.targetStep) {
+      return "partial";
+    } else {
+      return "discard";
+    }
+  }
+
+  private addResult(result: unknown, stepId?: number) {
+    this.addStep({
+      stepId: stepId ?? this.stepCount,
+      out: result,
+      concurrent: 1,
+      targetStep: 0,
+    });
+  }
+
+  private addStep(step: Step | Step[]) {
+    if (Array.isArray(step)) {
+      this.pendingSteps = [...this.pendingSteps, ...step];
+    } else {
+      this.pendingSteps.push(step);
+    }
+  }
+
+  private async invokeQstash(step: Step) {
+    const headers: Record<string, string> = {
+      [`Upstash-Forward-${internalHeader}`]: "yes",
+      "Upstash-Forward-Upstash-Workflow-Id": this.workflowId,
+      "Upstash-Workflow-Id": this.workflowId,
+    };
+
+    await this.client.publishJSON({
+      headers: headers,
+      method: "POST",
+      body: JSON.stringify(step),
+      url: this.url,
+      notBefore: step.sleepUntil,
+      delay: step.sleepFor,
+    });
+  }
+
+  private async sendPendingToQstash() {
+    // TODO: batch request for concurrent requests
+    for (const step of this.pendingSteps) {
+      await this.invokeQstash(step);
+    }
+    this.pendingSteps = [];
+  }
+
+  /**
+   * STATIC METHODS
+   */
+
+  static async parseRequest(request: Request) {
+    const callHeader = request.headers.get(internalHeader);
+    const firstCall = !callHeader;
```

Comment:
I think instead of call we can rely on `step` for better readability. `isFirstStep`.

Reply:
Would `isFirstInvocation` be nicer? because isFirstStep could refer to something else: first step in the workflow

Reply:
Much better than `step`.